### PR TITLE
✨(frontend) add page organization download contracts

### PR DESCRIPTION
**Pull request by rlecellier**
:warning: This PR is rebased on https://github.com/openfun/richie/pull/2091, do not review the first commit about bulk renames.

## Purpose

Add a page for organizations to signe they contracts, see issue: https://github.com/openfun/richie/issues/2075

### CODE REVIEW

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+import { useContracts } from 'hooks/useContacts';
+import Banner, { BannerType } from 'components/Banner';
+import { PER_PAGE } from 'settings';
+import { ContractFilters } from 'types/Joanie';
+
+const messages = defineMessages({
+  errorNoCourseProductRelation: {
+    defaultMessage: "This product doesn't exist",
+    description: 'Message displayed when requested course product relation is not found',
+    id: 'components.TeacherDashboardContracts.errorNoCourseProductRelation',
```
Comment (PanchoutNathan) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx:14]:
```suggestion
    id: 'pages.TeacherDashboardContractsLoader.TeacherDashboardContracts.errorNoCourseProductRelation'
```

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    },
+  ]);
+
+  const pagination = usePagination({
+    defaultPage: page ? parseInt(page, 10) : 1,
+    pageSize: PER_PAGE.teacherContractList,
+  });
+  useEffect(() => {
+    if (isFetched) {
+      pagination.setPagesCount(Math.ceil(meta!.pagination!.count / PER_PAGE.teacherContractList));
```
Comment (PanchoutNathan) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx:70]: Why not use contracts.length? I understand that it's exactly the same thing, but as you use contracts in the dependencies, it seems easier to understand to use contract.length. Or in this case, change the dependency of the useEffect

Comment (PanchoutNathan) [reply src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx:70]:
```
useEffect(() => {
    if (isFetched) {
      pagination.setPagesCount(Math.ceil(contracts.length / PER_PAGE.teacherContractList));
    }
  }, [contracts, isFetched]);
```

or 

```
useEffect(() => {
    if (isFetched) {
       pagination.setPagesCount(Math.ceil(meta!.pagination!.count / PER_PAGE.teacherContractList));
    }
  }, [meta.pagination.count, isFetched]);
```

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardCourseContractsLoader/index.tsx [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import { DashboardLayout } from 'widgets/Dashboard/components/DashboardLayout';
+import useQueryParams from 'hooks/useQueryParams';
+import TeacherDashboardContracts from '../TeacherDashboardContracts';
+import ContractFilters, { ContractListFilters } from '../ContractFilters';
+
+const messages = defineMessages({
+  pageTitle: {
+    defaultMessage: 'Contracts',
+    description: "Use for the page title of the course's contracts page",
+    id: 'components.TeacherDashboardContracts.pageTitle',
```
Comment (PanchoutNathan) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardCourseContractsLoader/index.tsx:15]: Ditto

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardCourseContractsLoader/index.tsx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  },
+  loading: {
+    defaultMessage: 'Loading contracts...',
+    description: 'Message displayed while loading contracts',
+    id: 'components.TeacherDashboardContracts.loading',
```
Comment (PanchoutNathan) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardCourseContractsLoader/index.tsx:20]: Ditto

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardOrganizationContractsLoader/index.tsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import { DashboardLayout } from 'widgets/Dashboard/components/DashboardLayout';
+import { TeacherDashboardOrganizationSidebar } from 'widgets/Dashboard/components/TeacherDashboardOrganizationSidebar';
+import useQueryParams from 'hooks/useQueryParams';
+import TeacherDashboardContracts from '../TeacherDashboardContracts';
+
+const messages = defineMessages({
+  pageTitle: {
+    defaultMessage: 'Contracts',
+    description: 'Use for the page title of the organization contracts area',
+    id: 'components.TeacherDashboardContracts.pageTitle',
```
Comment (PanchoutNathan) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardOrganizationContractsLoader/index.tsx:13]: Ditto

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardOrganizationContractsLoader/index.tsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  },
+  loading: {
+    defaultMessage: 'Loading contracts...',
+    description: 'Message displayed while loading contracts',
+    id: 'components.TeacherDashboardContracts.loading',
```
Comment (PanchoutNathan) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardOrganizationContractsLoader/index.tsx:18]: Ditto

File: src/frontend/js/pages/TeacherDashboardContractsLoader/index.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+export { TeacherDashboardCourseContractsLoader } from './TeacherDashboardCourseContractsLoader';
+export { TeacherDashboardOrganizationContractsLoader } from './TeacherDashboardOrganizationContractsLoader';
+
+// export TeacherDashboardCourseContractsLoader, TeacherDashboardOrganizationContractsLoader
```
Comment (PanchoutNathan) [inline src/frontend/js/pages/TeacherDashboardContractsLoader/index.ts:4]: Oopsi

File: src/frontend/js/widgets/Dashboard/components/TeacherDashboardCourseSidebar/utils.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+      TeacherDashboardPaths.ORGANIZATION_COURSE_GENERAL_INFORMATION,
+      TeacherDashboardPaths.ORGANIZATION_COURSE_CONTRACTS,
+    ];
+  }
+
+  if (courseProductRelationId) {
+    return [TeacherDashboardPaths.COURSE_PRODUCT, TeacherDashboardPaths.COURSE_PRODUCT_CONTRACTS];
+  }
+  return [TeacherDashboardPaths.COURSE_GENERAL_INFORMATION, TeacherDashboardPaths.COURSE_CONTRACTS];
+};
```
Comment (PanchoutNathan) [inline src/frontend/js/widgets/Dashboard/components/TeacherDashboardCourseSidebar/utils.ts:25]:
```suggestion 
return courseProductRelationId
    ? [TeacherDashboardPaths.COURSE_PRODUCT, TeacherDashboardPaths.COURSE_PRODUCT_CONTRACTS]
    : [TeacherDashboardPaths.COURSE_GENERAL_INFORMATION, TeacherDashboardPaths.COURSE_CONTRACTS];
```

File: src/frontend/js/widgets/Dashboard/components/TeacherDashboardCourseSidebar/utils.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    if (courseProductRelationId) {
+      return [
+        TeacherDashboardPaths.ORGANIZATION_PRODUCT,
+        TeacherDashboardPaths.ORGANIZATION_PRODUCT_CONTRACTS,
+      ];
+    }
+    return [
+      TeacherDashboardPaths.ORGANIZATION_COURSE_GENERAL_INFORMATION,
+      TeacherDashboardPaths.ORGANIZATION_COURSE_CONTRACTS,
+    ];
```
Comment (PanchoutNathan) [inline src/frontend/js/widgets/Dashboard/components/TeacherDashboardCourseSidebar/utils.ts:18]:
```suggestion
return courseProductRelationId
      ? [
          TeacherDashboardPaths.ORGANIZATION_PRODUCT,
          TeacherDashboardPaths.ORGANIZATION_PRODUCT_CONTRACTS,
        ]
      : [
          TeacherDashboardPaths.ORGANIZATION_COURSE_GENERAL_INFORMATION,
          TeacherDashboardPaths.ORGANIZATION_COURSE_CONTRACTS,
        ];
```

File: src/frontend/js/hooks/useQueryParams/index.ts [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import queryString from 'query-string';
+import { useMemo } from 'react';
+import { useLocation } from 'react-router-dom';
+
+const useQueryParams = <T extends {}>() => {
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply src/frontend/js/hooks/useQueryParams/index.ts:5]: I felt that it was too simple to have a test about it but i'll add a test.

File: src/frontend/js/pages/TeacherDashboardContractsLoader/ContractFilters/index.tsx [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+  onFiltersChange: (filters: ContractListFilters) => void;
+}
+
+const ContractFilters = ({ onFiltersChange }: ContractFiltersProps) => {
+  const intl = useIntl();
+  const {
+    items: organizations,
+    states: { isFetched },
+  } = useOrganizations();
+  const organizationOptions = organizations.map((organization) => ({
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply src/frontend/js/pages/TeacherDashboardContractsLoader/ContractFilters/index.tsx:27]: i didn't know that we had guidelines about blank lines.

File: src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx (hunk 2 above)
Comment (rlecellier) [reply src/frontend/js/pages/TeacherDashboardContractsLoader/TeacherDashboardContracts/index.tsx:70]: `contracts.lenght` will be 25 or less (the pagination page size).
Here we want the total number of page, this information is in `pagination.count`

File: src/frontend/js/utils/test/factories/joanie.ts [hunk 12]
```diff
@@ -68,6 +69,16 @@ export const TargetCourseFactory = factory((): TargetCourse => {
   };
 });
 
+export const ContractFactory = factory((): Contract => {
+  const uuid = faker.string.uuid();
+  return {
+    id: uuid,
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply src/frontend/js/utils/test/factories/joanie.ts:75]: copy/past, i'll change it.

File: src/frontend/js/widgets/Dashboard/components/TeacherDashboardCourseSidebar/utils.ts (hunk 9 above)
Comment (rlecellier) [reply src/frontend/js/widgets/Dashboard/components/TeacherDashboardCourseSidebar/utils.ts:18]: i think the early return is easier to read.

File: src/frontend/mocks/handlers.ts [hunk 13]
```diff
@@ -1 +1,3 @@
-export const handlers = [];
+import { getContracts } from './handlers/contracts';
+
+export const handlers = [getContracts];
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply src/frontend/mocks/handlers.ts:3]: my idea was to list here every handled routes like :
```js
getContracts,
downloadContract,
getOrders,
// etc
```
With an array we'll have:
* one less file to update when adding a new endpoint in contracts.
* to open every handler files (ex: contracts.ts, orders.ts, etc) to get the full list of mocked endpoints.
Both single exports and array export have pro/cond.

File: src/frontend/js/api/joanie.ts [hunk 14]
```diff
@@ -352,6 +355,21 @@ const API = (): Joanie.API => {
...
+      },
+      contracts: {
+        get: async (filters) => {
+          let url;
+          const { id = '', ...queryParameters } = filters || {};
+
+          if (id) url = ROUTES.user.contracts.get.replace(':id', id);
+          else url = ROUTES.user.contracts.get.replace(':id/', '');
+
+          if (!ObjectHelper.isEmpty(queryParameters)) {
```
[the remark this answers is not in the document]
Comment (rlecellier) [reply src/frontend/js/api/joanie.ts:369]: see https://github.com/openfun/richie/pull/2092

**Review by PanchoutNathan — APPROVED** [review]
LGTM

**Pull request merged**